src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,596 @@
+#include <Poco/Dynamic/Var.h>
+#include <Poco/DigestStream.h>
+#include <Poco/MD5Engine.h>
+#include <base/logger_useful.h>
+#include "Common/Exception.h"
+#include <fstream>
+#include <filesystem>
+#include <unistd.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int BAD_GET;
+    extern const int LOGICAL_ERROR;
+}
+
+std::shared_ptr<RemoteCacheController>
+RemoteCacheController::recover(const std::string & local_path_, std::function<void(RemoteCacheController *)> const & finish_callback)
+{
+    std::filesystem::path dir_handle(local_path_);
+    std::filesystem::path data_file(local_path_ + "/data.bin");
+    std::filesystem::path meta_file(local_path_ + "/meta.txt");
+    if (!std::filesystem::exists(dir_handle) || !std::filesystem::exists(data_file) || !std::filesystem::exists(meta_file))
+    {
+        LOG_ERROR(&Poco::Logger::get("RemoteCacheController"), "not exists directory:" + local_path_);
+        return nullptr;
+    }
+
+    std::ifstream meta_fs(local_path_ + "/meta.txt");
+    Poco::JSON::Parser meta_parser;
+    auto meta_jobj = meta_parser.parse(meta_fs).extract<Poco::JSON::Object::Ptr>();
+    auto remote_path = meta_jobj->get("remote_path").convert<std::string>();
+    auto schema = meta_jobj->get("schema").convert<std::string>();
+    auto cluster = meta_jobj->get("cluster").convert<std::string>();
+    auto downloaded = meta_jobj->get("downloaded").convert<std::string>();
+    auto mod_ts = meta_jobj->get("last_modification_timestamp").convert<UInt64>();
+    if (downloaded == "false")
+    {
+        LOG_ERROR(&Poco::Logger::get("RemoteCacheController"), "not a downloaded file: " + local_path_);
+        return nullptr;
+    }
+    auto size = std::filesystem::file_size(data_file);
+
+    auto cntrl = std::make_shared<RemoteCacheController>(schema, cluster, remote_path, mod_ts, local_path_, 0, nullptr, finish_callback);
+    cntrl->download_finished = true;
+    cntrl->current_offset = size;
+    meta_fs.close();
+
+    finish_callback(cntrl.get());
+    return cntrl;
+}
+
+RemoteCacheController::RemoteCacheController(
+    const std::string & schema_,
+    const std::string & cluster_,
+    const std::string & path_,
+    UInt64 ts_,
+    const std::string & local_path_,
+    size_t cache_bytes_before_flush_,
+    std::shared_ptr<ReadBuffer> readbuffer_,
+    std::function<void(RemoteCacheController *)> const & finish_callback)
+{
+    download_thread = nullptr;
+    schema = schema_;
+    cluster = cluster_;
+    local_path = local_path_;
+    remote_path = path_;
+    last_modification_timestamp = ts_;
+    local_cache_bytes_read_before_flush = cache_bytes_before_flush_;
+    valid = true;
+    if (readbuffer_ != nullptr)
+    {
+        download_finished = false;
+        current_offset = 0;
+        remote_readbuffer = readbuffer_;
+        // setup local files
+        out_file = new std::ofstream(local_path_ + "/data.bin", std::ios::out | std::ios::binary);
```

Comment:
Better to use `WriteBufferFromFile`, and if binary then `writeBinary` from `writeHelpers.h`.

Reply:
It may cause unnecessary memory copy if use WriteBufferFromFile and ReadBufferFromFile